Insights/EventsController.swift

```diff
@@ -0,0 +1,124 @@
+//
+//  EventsController.swift
+//  Insights
+//
+//  Created by Vladislav Fitc on 06/11/2018.
+//  Copyright © 2018 Algolia. All rights reserved.
+//
+
+import Foundation
+
+class EventsController: EventProcessor {
+    
+    typealias Storage = LocalStorage<[EventsPackage]>
+    
+    let credentials: Credentials
+    var eventsPackages: [EventsPackage]
+    let webservice: WebService
+    let logger: Logger
+    var isLocalStorageEnabled: Bool = true {
+        didSet {
+            if !isLocalStorageEnabled {
+                Storage.deleteFile(atPath: localStorageFileName)
+            }
+        }
+    }
+    private let dispatchQueue: DispatchQueue
+    private let localStorageFileName: String
+    private weak var flushTimer: Timer?
+    
+    init(credentials: Credentials,
+         webService: WebService,
+         flushDelay: TimeInterval,
+         logger: Logger,
+         dispatchQueue: DispatchQueue = .init(label: "insights.events", qos: .background)) {
+        self.eventsPackages = []
+        self.credentials = credentials
+        self.logger = logger
+        self.webservice = webService
+        self.dispatchQueue = dispatchQueue
+        self.localStorageFileName = "\(credentials.appId).events"
+        let flushTimer = Timer.scheduledTimer(timeInterval: flushDelay,
+                                                    target: self,
+                                                  selector: #selector(flushEvents),
+                                                  userInfo: nil,
+                                                   repeats: true)
+        self.flushTimer = flushTimer
+
+        RunLoop.main.add(flushTimer, forMode: .default)
+        deserialize()
+    }
+    
+    func process(_ event: Event) {
+        dispatchQueue.async { [weak self] in
+            self?.syncProcess(event)
+        }
+    }
+    
+    private func syncProcess(_ event: Event) {
+        let wrappedEvent = EventWrapper(event)
+        let eventsPackage: EventsPackage
+        
+        if let lastEventsPackage = eventsPackages.last, !lastEventsPackage.isFull {
+            // We are sure that try! will not crash, as where is "not full" check
+            eventsPackage = try! eventsPackages.removeLast().appending(wrappedEvent)
+        } else {
+            eventsPackage = EventsPackage(event: wrappedEvent)
+        }
+        
+        eventsPackages.append(eventsPackage)
+        serialize()
+    }
+    
+    @objc private func flushEvents() {
+        flush(eventsPackages)
+    }
+    
+    func flush(_ eventsPackages: [EventsPackage]) {
```

Comment:
for consistency with the method sync below, can we have named parameter for this method as well?